File: src/cash.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { initial } from "./game.js";
import { featureStep } from "./features.js";
test("cash cap blocks new coins only after both multiplayer teams reach it", () => {
 const s=initial();for(const p of s.players)p.stun=100;
 const item=s.pickups[2];s.credits=[2000,1900];item.wait=.04;
 featureStep(s,.04);assert.equal(item.wait,0);
 s.credits[1]=2000;item.wait=.04;
 featureStep(s,.04);assert.equal(item.wait,256/25);
 // Existing coins are still available; this is not a clamp on earnings.
 const p=s.players[7];Object.assign(p,{x:item.x,z:item.z,stun:0,action:0});item.wait=0;
 featureStep(s,.04);assert.equal(s.credits[0],2100);assert.ok(item.wait>0);
});
test("training uses the human cash threshold, independent of AI earnings", () => {
 const s=initial({training:true});for(const p of s.players)p.stun=100;
 assert.deepEqual(s.cashLimits,[10000,10000]);
 const item=s.pickups[2];s.credits=[9900,20000];item.wait=.04;
 featureStep(s,.04);assert.equal(item.wait,0);
 s.credits=[10000,0];item.wait=.04;
 featureStep(s,.04);assert.equal(item.wait,256/25);
});

File: server/cash_test.go
package main

import "testing"

func TestCashSpawnLimit(t *testing.T) {
	s := initial()
	for i := range s.Players {
		s.Players[i].Stun = 100
	}
	item := &s.Pickups[2]
	s.Credits = [2]int{2000, 1900}
	item.Wait = .04
	s.featureStep(.04)
	if item.Wait != 0 {
		t.Fatal("other team below limit")
	}
	s.Credits[1] = 2000
	item.Wait = .04
	s.featureStep(.04)
	if item.Wait != 256./25 {
		t.Fatal("both at limit")
	}
	p := &s.Players[7]
	p.X = item.X
	p.Z = item.Z
	p.Stun = 0
	p.Action = 0
	item.Wait = 0
	s.featureStep(.04)
	if s.Credits[0] != 2100 || item.Wait <= 0 {
		t.Fatal("visible coins remain collectible")
	}
}
func TestTrainingCashLimit(t *testing.T) {
	s := initial()
	s.Training = true
	s.initFeatures()
	for i := range s.Players {
		s.Players[i].Stun = 100
	}
	if s.CashLimits != [2]int{10000, 10000} {
		t.Fatal("practice limit")
	}
	item := &s.Pickups[2]
	s.Credits = [2]int{9900, 20000}
	item.Wait = .04
	s.featureStep(.04)
	if item.Wait != 0 {
		t.Fatal("human below limit")
	}
	s.Credits = [2]int{10000, 0}
	item.Wait = .04
	s.featureStep(.04)
	if item.Wait != 256./25 {
		t.Fatal("human at limit")
	}
}
